### PR TITLE
RES-863:  Use 'Layer Norm' instead of 'Batch Norm' on linear networks

### PR DESCRIPTION
**Pull request by lscheinkman**
@subutai Please review.
The main change in this PR is the switch from `BatchNorm` to `LayerNorm` on `LinearSDR` module.
See https://arxiv.org/abs/1607.06450

### CODE REVIEW

**Comment by subutai** [comment]
This doesn’t affect any of our existing sparse networks right?

**Comment by lscheinkman** [comment]
If affects `LinearSDR`. We can now use batch_size=1

**Comment by lscheinkman** [comment]
However we cannot user `LayerNorm` on `CNNSDR2d`. It is only good for Linear and RNN networks

File: htmresearch/frameworks/pytorch/cnn_sdr.py [hunk 1]
```diff
@@ -124,6 +129,8 @@ def __init__(self,
     shape = self.outputSize()
```
Comment (subutai) [inline htmresearch/frameworks/pytorch/cnn_sdr.py]: Don't you have to change outputSize() too?

File: htmresearch/frameworks/pytorch/linear_sdr.py [hunk 2]
```diff
@@ -81,10 +81,12 @@ def __init__(self,
     self.l1 = nn.Linear(inputFeatures, self.n)
     self.weightSparsity = weightSparsity
     self.learningIterations = 0
+    if self.k <= 0:
+      self.k = self.n
 
     self.bn = None
     if useBatchNorm:
-      self.bn = nn.BatchNorm1d(self.n, affine=False)
+      self.bn = nn.LayerNorm(self.n)
```
Comment (subutai) [inline htmresearch/frameworks/pytorch/linear_sdr.py]: How would this affect GSC results?

File: htmresearch/frameworks/pytorch/linear_sdr.py [hunk 3]
```diff
@@ -81,10 +81,12 @@ def __init__(self,
     self.l1 = nn.Linear(inputFeatures, self.n)
     self.weightSparsity = weightSparsity
     self.learningIterations = 0
+    if self.k <= 0:
+      self.k = self.n
 
     self.bn = None
     if useBatchNorm:
```
Comment (subutai) [inline htmresearch/frameworks/pytorch/linear_sdr.py]: Should probably change the name of the parameter too.

File: htmresearch/frameworks/pytorch/cnn_sdr.py (hunk 1 above)
Comment (lscheinkman) [reply htmresearch/frameworks/pytorch/cnn_sdr.py]: `outputSize` was already using stride parameter
